File: apps/server/src/modules/orders/orders.service.ts
import { OrderStatus, PackageSize } from '@prisma/client';
import { prisma } from '../../lib/prisma';
import { AppError } from '../../middleware/error.middleware';
import { broadcastOrderUpdate } from '../../lib/websocket';

let orderCounter = 0;

async function getNextOrderId(): Promise<string> {
  if (orderCounter === 0) {
    const latest = await prisma.order.findFirst({ orderBy: { createdAt: 'desc' } });
    const match = latest?.orderId.match(/#BZ-(\d+)/);
    orderCounter = match ? parseInt(match[1]) : 2400;
  }
  return `#BZ-${++orderCounter}`;
}

export interface CreateOrderDto {
  recipientName: string;
  recipientPhone: string;
  address: string;
  area: string;
  areaAr?: string;
  governorate: string;
  governorateAr?: string;
  packageSize?: PackageSize;
  isFragile?: boolean;
  isCod?: boolean;
  codAmount?: number;
  notes?: string;
}

export interface ListOrdersDto {
  status?: OrderStatus;
  search?: string;
  page?: number;
  limit?: number;
  from?: string;
  to?: string;
}

export async function createOrder(merchantId: string, dto: CreateOrderDto) {
  const orderId = await getNextOrderId();
  const order = await prisma.order.create({
    data: {
      orderId,
      merchantId,
      ...dto,
      packageSize: dto.packageSize ?? PackageSize.MEDIUM,
      status: OrderStatus.PENDING,
    },
    include: { statusHistory: true },
  });

  await prisma.orderStatusHistory.create({
    data: { orderId: order.id, status: OrderStatus.PENDING, note: 'Order created' },
  });

  broadcastOrderUpdate(merchantId, order);
  return order;
}

export async function listOrders(merchantId: string, dto: ListOrdersDto) {
  const page = dto.page ?? 1;
  const limit = Math.min(dto.limit ?? 20, 100);
  const skip = (page - 1) * limit;

  const where = {
    merchantId,
    ...(dto.status && { status: dto.status }),
    ...(dto.from || dto.to
      ? {
          createdAt: {
            ...(dto.from && { gte: new Date(dto.from) }),
            ...(dto.to && { lte: new Date(dto.to) }),
          },
        }
      : {}),
    ...(dto.search && {
      OR: [
        { orderId: { contains: dto.search, mode: 'insensitive' as const } },
        { recipientName: { contains: dto.search, mode: 'insensitive' as const } },
        { area: { contains: dto.search, mode: 'insensitive' as const } },
      ],
    }),
  };

  const [orders, total] = await prisma.$transaction([
    prisma.order.findMany({
      where,
      orderBy: { createdAt: 'desc' },
      skip,
      take: limit,
      include: { statusHistory: { orderBy: { createdAt: 'asc' } } },
    }),
    prisma.order.count({ where }),
  ]);

  return {
    data: orders,
    meta: { page, limit, total, totalPages: Math.ceil(total / limit) },
  };
}

export async function getOrder(merchantId: string, id: string) {
  const order = await prisma.order.findFirst({
    where: { OR: [{ id }, { orderId: id }], merchantId },
    include: { statusHistory: { orderBy: { createdAt: 'asc' } } },
  });
  if (!order) throw new AppError(404, 'Order not found');
  return order;
}

export async function updateOrderStatus(
  merchantId: string,
  id: string,
  status: OrderStatus,
  note?: string,
) {
  const order = await getOrder(merchantId, id);

  const updated = await prisma.order.update({
    where: { id: order.id },
    data: { status },
    include: { statusHistory: { orderBy: { createdAt: 'asc' } } },
  });

  await prisma.orderStatusHistory.create({
    data: { orderId: order.id, status, note },
  });

  broadcastOrderUpdate(merchantId, updated);
  return updated;
}

export async function cancelOrder(merchantId: string, id: string) {
  const order = await getOrder(merchantId, id);
  const nonCancellable: OrderStatus[] = [OrderStatus.DELIVERED, OrderStatus.CANCELLED];
  if (nonCancellable.includes(order.status)) {
    throw new AppError(400, `Cannot cancel an order with status ${order.status}`);
  }
  return updateOrderStatus(merchantId, id, OrderStatus.CANCELLED, 'Cancelled by merchant');
}

File: apps/mobile/lib/services/orders_service.dart
import 'api_client.dart';

enum OrderStatus { pending, processing, inDelivery, delivered, cancelled }

extension OrderStatusX on OrderStatus {
  String get apiValue => switch (this) {
        OrderStatus.pending => 'PENDING',
        OrderStatus.processing => 'PROCESSING',
        OrderStatus.inDelivery => 'IN_DELIVERY',
        OrderStatus.delivered => 'DELIVERED',
        OrderStatus.cancelled => 'CANCELLED',
      };

  static OrderStatus fromApi(String s) => switch (s) {
        'PENDING' => OrderStatus.pending,
        'PROCESSING' => OrderStatus.processing,
        'IN_DELIVERY' => OrderStatus.inDelivery,
        'DELIVERED' => OrderStatus.delivered,
        'CANCELLED' => OrderStatus.cancelled,
        _ => OrderStatus.pending,
      };

  String label(bool isAr) => switch (this) {
        OrderStatus.pending => isAr ? 'قيد الانتظار' : 'Pending',
        OrderStatus.processing => isAr ? 'قيد المعالجة' : 'Processing',
        OrderStatus.inDelivery => isAr ? 'في الطريق' : 'In Delivery',
        OrderStatus.delivered => isAr ? 'تم التوصيل' : 'Delivered',
        OrderStatus.cancelled => isAr ? 'ملغي' : 'Cancelled',
      };
}

class Order {
  final String id;
  final String orderId;
  final String recipientName;
  final String recipientPhone;
  final String address;
  final String area;
  final String? areaAr;
  final String governorate;
  final String? governorateAr;
  final OrderStatus status;
  final String? driverName;
  final String? driverNameAr;
  final String? driverPhone;
  final bool isCod;
  final double codAmount;
  final bool isFragile;
  final String? notes;
  final DateTime createdAt;

  const Order({
    required this.id,
    required this.orderId,
    required this.recipientName,
    required this.recipientPhone,
    required this.address,
    required this.area,
    this.areaAr,
    required this.governorate,
    this.governorateAr,
    required this.status,
    this.driverName,
    this.driverNameAr,
    this.driverPhone,
    required this.isCod,
    required this.codAmount,
    required this.isFragile,
    this.notes,
    required this.createdAt,
  });

  factory Order.fromJson(Map<String, dynamic> j) => Order(
        id: j['id'] as String,
        orderId: j['orderId'] as String,
        recipientName: j['recipientName'] as String,
        recipientPhone: j['recipientPhone'] as String,
        address: j['address'] as String,
        area: j['area'] as String,
        areaAr: j['areaAr'] as String?,
        governorate: j['governorate'] as String,
        governorateAr: j['governorateAr'] as String?,
        status: OrderStatusX.fromApi(j['status'] as String),
        driverName: j['driverName'] as String?,
        driverNameAr: j['driverNameAr'] as String?,
        driverPhone: j['driverPhone'] as String?,
        isCod: j['isCod'] as bool? ?? false,
        codAmount: (j['codAmount'] as num?)?.toDouble() ?? 0,
        isFragile: j['isFragile'] as bool? ?? false,
        notes: j['notes'] as String?,
        createdAt: DateTime.parse(j['createdAt'] as String),
      );
}

class OrdersResult {
  final List<Order> data;
  final int total;
  final int page;
  final int totalPages;

  const OrdersResult({
    required this.data,
    required this.total,
    required this.page,
    required this.totalPages,
  });
}

class OrdersService {
  OrdersService._();
  static final OrdersService instance = OrdersService._();

  final _api = ApiClient.instance;

  Future<OrdersResult> listOrders({
    OrderStatus? status,
    String? search,
    int page = 1,
    int limit = 20,
  }) async {
    final params = <String, String>{
      'page': '$page',
      'limit': '$limit',
      if (status != null) 'status': status.apiValue,
      if (search != null && search.isNotEmpty) 'search': search,
    };
    final query = params.entries.map((e) => '${e.key}=${e.value}').join('&');
    final res = await _api.get('/orders?$query');
    final list = (res['data'] as List).map((e) => Order.fromJson(e as Map<String, dynamic>)).toList();
    final meta = res['meta'] as Map<String, dynamic>;
    return OrdersResult(
      data: list,
      total: meta['total'] as int,
      page: meta['page'] as int,
      totalPages: meta['totalPages'] as int,
    );
  }

  Future<Order> getOrder(String id) async {
    final res = await _api.get('/orders/$id');
    return Order.fromJson(res);
  }

  Future<Order> createOrder(Map<String, dynamic> dto) async {
    final res = await _api.post('/orders', dto);
    return Order.fromJson(res);
  }

  Future<Order> cancelOrder(String id) async {
    final res = await _api.delete('/orders/$id');
    return Order.fromJson(res);
  }
}
